### PR TITLE
Refactor to allow halo > 0 decomposition of the mesh

### PR DESCRIPTION
**Pull request by twsearle**
## Description

Split the mesh partitioner into components:
 1.	Generation of the orca global grid (blue box). This is already performed by the OrcaGrid object.
 2.	Generation of the matching global regular lat-lon grid (orange box). This is a simple atlas RegularLonLatGrid structure. This is partitioned by the chosen partitioner to generate a distribution, which we use to determine the partition of each node in subsequent steps
 3.	A local `SurroundingRectangle` (better name might be LocalLonLatGrid, green box) that surrounds the points held by this partition.
 4.	A `LocalOrcaGrid` object that includes any orca periodic points at the edges of the domain owned by the partition.

image

This allows us to set the halo at the generation of the surrounding rectangle, and separate out the logic to handle the strangeness of the problems at the orca grid edge.

Note, there were some strange rogue points in the lower portion of the grid that were not seemingly correctly marked as ghost points. This has had no impact to correct, other than it affects our calculation of field norms in orca-jedi.

## Issues 

- ecmwf/atlas-orca/issues/19

## Acceptance criteria

  * All ctests pass
  * halo = 1 configuration of JEDI framework executable properly locates observations on the correct partition.
  * performance after the change is comparable with performance before the change.

## Testing

  * ctests added to cover the components
  * Integration testing as part of [orca-jedi ](https://github.com/MetOffice/orca-jedi/pull/92)
  * Regression testing as part of a full comparison workflow at the Met Office via [jjdocs change](https://github.com/MetOffice/jjdocs/pull/167)

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/atlas-orca?pullRequest=20) <br/>All committers have signed the CLA.

**Review by wdeconinck — COMMENTED** [review]
I think splitting these parts up is a very good idea @twsearle thanks for taking this up!

I applied a few fixes to your branch related to normalisation and IJ-indexing into the north-fold halo.

File: src/atlas-orca/meshgenerator/LocalOrcaGrid.cc
Comment (twsearle) [inline src/atlas-orca/meshgenerator/LocalOrcaGrid.cc:243]: I don't think I understand this one (sorry its been a while since we made this change and I am trying to remind myself of the context).

Firstly its different to the same limit in `normalised_grid_master_lonlat` where previously these two cases were the same. whats the reason for that? Secondly, this change uses `ix` rather than the master coordinate. I think in the original code this used the master coordinate:
https://github.com/ecmwf/atlas-orca/blob/d9ff4d39ff3e7a420548a207fef60969c2bbf4f4/src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc#L396

File: src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc
Comment (twsearle) [inline src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: Is it possible to express the difference in this change in terms of some sort of named coordinate? Sorry my head always start to spin as soon as I look at this mesh generator again - I was hoping I had captured all the cases in terms of functions (although not efficient we were going for clarity first because we had so many issues getting it correct for higher res grids). 

`orca_haloed_global_grid_ij` is supposed to wrap points if they extend outside the original orca halo in the case of mpi with halo > 0, but behave as you have it here when inside that region.

It could be that the case I was catching no longer happens now that the code is setup as I have it, but in that case I think what you want is `global_ij` because you are not concerned about this wrapping issue?

**Comment by twsearle** [comment]
> I think splitting these parts up is a very good idea @twsearle thanks for taking this up!
> 
> I applied a few fixes to your branch related to normalisation and IJ-indexing into the north-fold halo.

I have done my best to understand the changes - apologies if I am misunderstanding this stuff is quite confusing and I made the changes over quite a large block of time - some of the functions/choices I made might be redundant.

I will rerun this version to double check it gives us the same results in our integration testing. During the process we thought we were done with this change and then found a whole extra series of issues with higher resolution grids.

Its a real slog to get through this review thanks so much for your time! Let me know if there is anything else I can do.

File: src/atlas-orca/meshgenerator/LocalOrcaGrid.cc
Comment (wdeconinck) [reply src/atlas-orca/meshgenerator/LocalOrcaGrid.cc:243]: My understanding is that `ix_orca_min_ + ix == ix_glb_master`, so that we are normalising with respect to the global left boundary. 
You can verify with following commands:
```bash
atlas-meshgen ORCA1_F --coordinates=lonlat --ghost --info
gmsh mesh.msh mesh_info.msh &
```
and look at the "water" field (or land-sea-mask):
With correction [d87a650](https://github.com/ecmwf/atlas-orca/pull/20/commits/d87a65070525e71a12c4822c50715a3ca3e7e6c5) and also before this PR:  
image
Before this correction: 
image
There is nothing wrong with those coordinates as the transformation to 3D XYZ is OK, but they are normalised such that with Gmsh we cannot see the picture clearly in lonlat coordinates.

File: src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc
Comment (s-good) [reply src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: I don't think the proposed change will work if halo size > 0. If halo_size > 0,  ix+ local_orca.ix_min() or iy + local_orca.iy_min() can be a point that is outside the bounds of the ORCA grid. By using local_orca.orca_haloed_global_grid_ij, any points that lie outside the ORCA grid bounds get wrapped to the correct points on the ORCA grid.

Comment (wdeconinck) [reply src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: The i,j coordinates are not needing to be wrapped. They are just there to see the grid structure.

To see the problem, we can look at the following command:
```bash
mpirun -np 2 bin/atlas-meshgen ORCA2_T --coordinates=ij --info --ghost --halo=1
gmsh mesh.msh mesh_info.msh &
```
and e.g. look again at the 'water' field.

With my modification [6509f04](https://github.com/ecmwf/atlas-orca/pull/20/commits/6509f048ad54c55e3e91dd81a986d827200fb7d5) applied:
image
Without the modification:
image

As you can see, without the modification there's no easy way to see the ghost region. With the modification and having i,j counting further into the ghost region it is easier to see what's going on.

Still it seems there are some issues in the corners even with the modification applied...

**Comment by wdeconinck** [comment]
It is definitely not an easy piece of code to understand and edit.
There are some problems still when looking at `halo` or `ghost` field in the halo (check mesh-info) I noticed.
However if you have your parallel algorithms working for you as is, then we could go ahead and solve that issue later..

File: src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc
Comment (wdeconinck) [reply src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: > I don't think the proposed change will work if halo size > 0. If halo_size > 0, ix+ local_orca.ix_min() or iy + local_orca.iy_min() can be a point that is outside the bounds of the ORCA grid. By using local_orca.orca_haloed_global_grid_ij, any points that lie outside the ORCA grid bounds get wrapped to the correct points on the ORCA grid.

It is OK to go outside the bounds of the original ORCA grid (up to any desired halo size) as long as it gets correctly marked as a periodic halo, and the halo-exchanges are done correctly. In the end what gets written to file is the grid without halo (or just the halos present from in the grid as was originally defined in netCDF format)

File: src/atlas-orca/meshgenerator/LocalOrcaGrid.cc
Comment (twsearle) [reply src/atlas-orca/meshgenerator/LocalOrcaGrid.cc:243]: Does the above hold with halo > 1 and mpi > 2? Sorry I am having a lot of trouble building atlas-orca at the moment at the Met Office (it will get easier once we all move to atlas 0.38.1) so it is tricky for me to check for myself right now.
 
I believe that the concept of the "master ij" changes its meaning a little when we introduce halos. We have been treating it as the ij point corresponding to the "master index" where this has the specific meaning of the index to the unique non-orca-halo point in the grid for that data. We do not use the word "master" for any other meaning.... at least that was the intention! :)

The point you are referring to is what we have called the `global_ij` point. Just the local ix, iy translated to global ij space with no wrapping.

I am not sure what it is we want, but I am very keen for us to use labelled coordinate systems. When doing this change I believe we counted about 7 coordinate systems in play - some of them reduce to others of them, but it is very confusing and the differences are often slight! This was my go at listing them off:

code | Indexing scheme for indexing 2D points into 1D arrays (i,j   meanings depend on the schema) | Methods/arrays using this scheme
-- | -- | --
Global regular | 0 indexing on the global regular grid (for indexing 1D array   regular grid objects) [j*nx_glb + i] | Atlas regular grid object, distribution.partition indexing
Local regular | 0   indexing on the local regular grid (for indexing 1D array properties of this   object) [j*nx + i] | is_node, is_ghost, partition array
Haloed local regular | Haloed indexing on local regular grid (starting from (-halo,   -halo)) | Surrounding rectangle
Global orca | 0   indexing on the global orca grid (for indexing 1D array OrcaGrid objects)   [j*orcagrid.nx() + i] | OrcaGrid   internal indexing
Folded global orca | Orcafold-Haloed indexing on the global orca grid (starting from   (-orca_halo_w, -orca_halo_s), used for indexing things via PeriodicIndex   function) [offset = -j_min *    orca_nx_halo – i_min; offset + j*orca_nx_halo + i] | OrcaGrid::PeriodicIndex to wrap around the orca grid and return   correct node.
Haloed local orca | Orcafold-haloed   indexing on the local orca grid (including the orca halos) [offset = -j_min   *  orca_nx_halo – i_min; offset +   j*orca_nx_halo + i] | Local Orca   Grid object when there are no mpi halos
Haloed   Local   Orca | Orcafold-and-mpi-haloed indexing on the local orca grid   (including both the orcafold points and the mpi haloed points) [offset =   -j_min *  orca_nx_halo – i_min; offset   + j*orca_nx_halo + i] | Local Orca Grid object when there are mpi halos
Local orca | 0   indexing on the local orca grid [j*nx_orca + i] | Local   Orcagrid is_node, is_ghost, partition array

Comment (twsearle) [reply src/atlas-orca/meshgenerator/LocalOrcaGrid.cc:243]: Its interesting how the water/land mask has become messed up too in the version we had. Honestly I kind of prefer the second version because it makes the cells in the north fold visible - it really does have that wacky shape. I can see it isn't playing nicely with visualisation though.

File: src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc
Comment (twsearle) [reply src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: > Still it seems there are some issues in the corners even with the modification applied...

This might be from my fixes :). In order to get those boxes to look all neat, you end up having ghost nodes that point to other ghost nodes in a previous implementation. This is very bad, because a lot of atlas is dereferencing ghost nodes one time to get to the master node in various places.

**Comment by twsearle** [comment]
> It is definitely not an easy piece of code to understand and edit. There are some problems still when looking at `halo` or `ghost` field in the halo (check mesh-info) I noticed. However if you have your parallel algorithms working for you as is, then we could go ahead and solve that issue later..

I _think_ everything is OK - I just checked and we are still passing our integration tests. Sorry for being super cautious about this one (I can see on reflection your changes only affect the fields used to view in gmsh) - I have thought I had it about a dozen times, made a minor change, and broken something somewhere! It has been rough. I think the dodgy looking cells are fundamentally the fault of the degeneracy in the orca grid as I mentioned above. I have regrets about choosing the ORCA2 grid for testing - its a very strange animal.

File: src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc
Comment (s-good) [reply src/atlas-orca/meshgenerator/OrcaMeshGenerator.cc:248]: > > I don't think the proposed change will work if halo size > 0. If halo_size > 0, ix+ local_orca.ix_min() or iy + local_orca.iy_min() can be a point that is outside the bounds of the ORCA grid. By using local_orca.orca_haloed_global_grid_ij, any points that lie outside the ORCA grid bounds get wrapped to the correct points on the ORCA grid.
> 
> It is OK to go outside the bounds of the original ORCA grid (up to any desired halo size) as long as it gets correctly marked as a periodic halo, and the halo-exchanges are done correctly. In the end what gets written to file is the grid without halo (or just the halos present from in the grid as was originally defined in netCDF format)

Thanks for this explanation and for fixing the code.

**Comment by twsearle** [comment]
@wdeconinck We are now using atlas v 0.38.1 🎉 in our stack! This means that we finally have everything in place to use this change once it is merged and tagged. Apologies again for my wordy comments/descriptions on this one, it really was rather painful! However we have been working with it for a while now and I feel much more confident with it now. Is there anyone else who needs to review it?

**Review by wdeconinck — APPROVED** [review]
Thanks @twsearle  for confirming all that! Good to merge then.

**Comment by twsearle** [comment]
Thanks very much for merging this @wdeconinck could we please get a new tag version including this change? I am excited to get it into our stack as soon as possible!

**Pull request merged**
